### PR TITLE
Country support improvements

### PR DESCRIPTION
**Pull request by FeralChild64**
# Description

- cleanup - moved all the lengthy/boring locale tables to `dos_locale.cpp`; file contains a little business code, but not much
- country settings now influence currency symbol/format and list separators
- added possibility to select between historic and modern locale
- added `--list-countries` command line switch
- changed default code page for Bulgaria, as requested by a native speaker @Torinde

Many thanks to @rderooy for his work on extracting locale data from various DOS'es.

#### Limitations / out of scope of this PR / not bugs

- there is no `--list-keyboard-layouts` or `--list-codepages`; I agree they could be useful, but there is not enough time to do this for 0.80.1, besides the values are documented in `contrib/resources/freedos-*` directories; the list of supported countries was not documented anywhere
- I agree that locale synchronization with host can be improved in many ways, but again - no time for this
- there are going to be slight differences between DOSBox and MS-DOS 6.22 regarding currency; the reason is that MS-DOS often does not care whether the current code page supports national characters needed for the currency (for example, country 48 and standard code page gives a borked currency symbol results in MS-DOS), while I do care for this; not a bug, but a feature :D
- when no country is set, MS-DOS 6.22 sets the US locale but with a different date separator; I believe this is a bug in MS-DOS, which is not replicated
- MS-DOS / PC-DOS / DR-DOS uses the same country code for _Australia_ and _International English_; I have decided to use 0 for _International English_ in our configuration file, but the DOS API will return the same country ID as for _Australia_, for possibly better compatibility with guest software (DOS seems to never return country code 0)
- collation and uppercase tables are not generated yet, defaults are used; especially collation table will require quite some work to be generated well
- our internal `dir` command always uses 4-digit year (in reality it should print 2-digit year in case for Historic locale) - I intend to fix this with an upcoming command rework (won't be ready for 0.81.0)

## Related issues

Fixes https://github.com/dosbox-staging/dosbox-staging/issues/3128

# Manual testing

[CTEST.C.gz](https://github.com/dosbox-staging/dosbox-staging/files/13536348/CTEST.C.gz)
[CTEST.EXE.gz](https://github.com/dosbox-staging/dosbox-staging/files/13536349/CTEST.EXE.gz)

Attached a locale test tool above. if run without parameters, it shows the current locale settings. If run with numeric parameter, it also tries to use DOS API to set the locale (note: at least on MS-DOS 5.0a, MS-DOS 6.22, PC-DOS 2000, and DR-DOS 7.03 it only allows to set the country which was specified in CONFIG.SYS file).

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [x] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by FeralChild64** [comment]
@rderooy The COUNTRY.SYS from OS/2 should be usable in MS-DOS, if I remember correctly, just like the one from Windows ME is.

OK, thanks for the review, but I want to give other reviewers a chance.

**Comment by rderooy** [comment]
I tried the COUNTRY.SYS from the Thai edition of MS-DOS 6.22, and it was identical to the English version.

File: src/dos/dos.cpp [hunk 1]
```diff
@@ -793,17 +654,27 @@ static Bitu DOS_21Handler(void) {
...
+			 // works when setting the country to the one currently
+			 // set - unknown, why
+			 countryNo = (reg_al == 0xff) ? reg_bx : reg_al;
+			 if (DOS_SetCountry(countryNo)) {
+				        reg_ax = 0;
+				        CALLBACK_SCF(false);
+			 } else {
+				        reg_ax = 0x02; // invalid country
+				        CALLBACK_SCF(true);
+			 }
```
Comment (johnnovak) [inline src/dos/dos.cpp:670]: reformat

File: src/dos/dos_keyboard_layout.cpp [hunk 2]
```diff
@@ -195,7 +192,9 @@ static uint32_t read_kcl_file(const FILE_unique_ptr &kcl_file, const char *layou
 
 	const auto seek_pos = 7 + rbuf[6];
 	if (fseek(kcl_file.get(), seek_pos, SEEK_SET) != 0) {
-		LOG_WARNING("LAYOUT: could not seek to byte %d in keyboard layout file: %s", seek_pos, strerror(errno));
+		LOG_WARNING("DOS: Could not seek to byte %d in keyboard layout file: %s",
```
Comment (johnnovak) [inline src/dos/dos_keyboard_layout.cpp:195]: That's a very specific error message. Does the use need to know the exact location of the byte? 😅 Does he care? 😆 Suggest changing it to something like `DOS: Error reading keyboard layout file: '%'`

File: src/dos/dos_keyboard_layout.cpp [hunk 3]
```diff
@@ -214,7 +213,7 @@ static uint32_t read_kcl_file(const FILE_unique_ptr &kcl_file, const char *layou
 
 		constexpr int8_t lang_codes_offset = -2;
 		if (fseek(kcl_file.get(), lang_codes_offset, SEEK_CUR) != 0) {
-			LOG_ERR("LAYOUT: could not seek to the language codes "
+			LOG_ERR("DOS: Could not seek to the language codes "
```
Comment (johnnovak) [inline src/dos/dos_keyboard_layout.cpp:216]: `DOS: Error reading keyboard layout file: '%'` ?

File: src/dos/dos_keyboard_layout.cpp [hunk 4]
```diff
@@ -1731,24 +1203,30 @@ class DOS_KeyboardLayout final : public Module_base {
 
 		if (loaded_layout->ReadKeyboardFile(layoutname.c_str(), dos.loaded_codepage)) {
 			if (strncmp(layoutname.c_str(), "auto", 4)) {
-				LOG_ERR("LAYOUT: Failed to load keyboard layout %s",
+				LOG_ERR("DOS: Failed to load keyboard layout %s",
```
Comment (johnnovak) [inline src/dos/dos_keyboard_layout.cpp:1206]: `: '%s'` (colon, apostrophes)

File: src/dos/dos_locale.cpp [hunk 5]
```diff
@@ -0,0 +1,2904 @@
...
+#endif
+
+#include "bitops.h"
+#include "checks.h"
+#include "dos_inc.h"
+#include "dos_keyboard_layout.h"
+#include "logging.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:39]: Noice 😎

File: src/dos/dos_locale.cpp [hunk 6]
```diff
@@ -0,0 +1,2904 @@
...
+
+// NOTE: Locale settings below were selected based on our knowledge and various
+// public sources. If in doubt, we choose what we think is relevant for the most
+// representative languange for the given country.
+// We do not mean to stigmatise any minorities, we do not mean to offend anyone,
+// we do our best to put our personal opinions and beliefs aside.
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:45]: A bit too touchy-feely for my taste, but if this preempts some drama, I guess some tiptoeing doesn't matter. You'd think no one would be upset about a slightly wrong localisation setting in an obscure DOS emulator, but nations went to war for less... 🥶 🤦🏻

File: src/dos/dos_locale.cpp [hunk 7]
```diff
@@ -0,0 +1,2904 @@
...
+// TODO: at least MS-DOS 6.22 and Windows ME unofficially support
+//       country code 711 - not sure what it represents:
+//       - date format  / separator             : YearMonthDay / '-'
+//       - time format / separator              : 24H / ':'
+//       - currency symbol / precision / format : 'EA$' / 2 / SymbolAmount
+//       - thousands /decimal / list separator  : '.' / ',' / ','
+
+// ***************************************************************************
+// Country info - time/date format, currency, etc.
+// ***************************************************************************
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:277]: Clearly, you've done your research 🥇

File: src/dos/dos_locale.cpp [hunk 8]
```diff
@@ -0,0 +1,2904 @@
...
+
+	// Current locale systems (like Unicode CLDR) do not seem to specify
+	// list separators anymore. On 'https://answers.microsoft.com' one can
+	// find a question:
+	// - Why does Excel seem to use ; and , differently per localization?
+	// And the top answer is:
+	// - In countries such as the USA and UK, the comma is used as list
+	//   separator. (...)
+	//   In countries that use comma as decimal separator, such as many
+	//   continental European countries, using a comma both as decimal
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:344]: "European countries, except for the UK" :trollface: (it's the truth)

File: src/dos/dos_locale.cpp [hunk 9]
```diff
@@ -0,0 +1,2904 @@
...
+	destination[offset + 0] = static_cast<uint8_t>(source.thousands_separator);
+	destination[offset + 1] = 0;
+
+	offset = InfoOffsetDecimalSeparator;
+	destination[offset + 0] = static_cast<uint8_t>(source.decimal_separator);
+	destination[offset + 1] = 0;
+
+	offset = InfoOffsetListSeparator;
+	destination[offset + 0] = static_cast<uint8_t>(source.list_separator);
+	destination[offset + 1] = 0;
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2025]: `enul_val`s instead of the casts?

File: src/dos/dos_locale.cpp [hunk 10]
```diff
@@ -0,0 +1,2904 @@
...
+		if (found) {
+			memcpy(&destination[InfoOffsetCurrencySymbol],
+			       candidate.c_str(),
+			       candidate.length());
+			break;
+		}
+	}
+
+	offset = InfoOffsetCurrencyFormat;
+	destination[offset] = static_cast<uint8_t>(source.currenct_format);
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2062]: `enum_val`?

File: src/dos/dos_locale.cpp [hunk 11]
```diff
@@ -0,0 +1,2904 @@
...
+	}
+
+	offset = InfoOffsetCurrencyPrecision;
+	destination[offset] = source.currency_precision;
+
+	// Mark locale as generated
+
+	config.is_locale_generated = true;
+
+	// If locale changed, log it
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2081]: Minor: I'd suggest to get rid of these extra blank lines after single line comments.

File: src/dos/dos_locale.cpp [hunk 12]
```diff
@@ -0,0 +1,2904 @@
...
+
+	static bool already_logged = false;
+
+	static Country logged_country            = {};
+	static LocalePeriod logged_real_period   = {};
+	static LocalePeriod logged_config_period = {};
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2087]: Can't we turn this into a class to get rid of the statics?

File: src/dos/dos_locale.cpp [hunk 13]
```diff
@@ -0,0 +1,2904 @@
...
+	LOG_MSG("DOS: Loaded %s locale for country %d, '%s'%s",
+	        period_to_string(locale_period),
+	        static_cast<int>(config.country),
+	        country_info.country_name.c_str(),
+	        additional_comment.c_str());
+
+	already_logged       = true;
+	logged_country       = config.country;
+	logged_real_period   = locale_period;
+	logged_config_period = config.locale_period;
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2119]: This is a huge function. Can't it be broken up into smaller logical chunks and/or lambdas with descriptive names and well-defined small scopes?

File: src/dos/dos_locale.cpp [hunk 14]
```diff
@@ -0,0 +1,2904 @@
...
+	if (!CountryData.count(country_corrected)) {
+		if (no_fallback) {
+			return false;
+		}
+
+		const auto default_country = static_cast<Country>(
+		        DOS_GetDefaultCountry());
+		LOG_WARNING("DOS: No locale info for country %d, using default %d",
+		            static_cast<int>(country),
+		            static_cast<int>(default_country));
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2140]: `enum_val`

File: src/dos/dos_locale.cpp [hunk 15]
```diff
@@ -0,0 +1,2904 @@
...
+		}
+	}
+	return language_code;
+}
+
+bool DOS_GetCountryFromLayout(const std::string& layout, uint16_t& country)
+{
+	const auto it = LayoutToCountryMap.find(layout);
+	if (it != LayoutToCountryMap.end()) {
+		country = static_cast<uint16_t>(it->second);
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2329]: `enum_val`

File: src/dos/dos_locale.cpp [hunk 16]
```diff
@@ -0,0 +1,2904 @@
...
+
+	case Country::Azerbaijan: return assert_code_page(58210);
+
+	case Country::Georgia: return assert_code_page(59829);
+
+	case Country::Uzbekistan: return assert_code_page(62306);
+
+	default:
+		LOG_WARNING("DOS: No default code page for country %d, '%s'",
+		            static_cast<int>(country_id),
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2479]: `enum_val`

File: src/dos/dos_locale.cpp [hunk 17]
```diff
@@ -0,0 +1,2904 @@
...
+	if (GetKeyboardLayoutName(layout_id_string) &&
+	    (safe_strlen(layout_id_string) == 8)) {
+		const int current_kb_layout_by_name = parse_hex_string(
+		        (char*)&layout_id_string[4]);
+		layout_id_string[4] = 0;
+
+		const int sub_id = parse_hex_string((char*)&layout_id_string[0]);
+
+		if ((current_kb_layout_by_name > 0) &&
+		    (current_kb_layout_by_name < 65536)) {
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2509]: `MAX_UINT16 + 1`?

File: src/dos/dos_locale.cpp [hunk 18]
```diff
@@ -0,0 +1,2904 @@
...
+}
+
+void DOS_Locale_AddMessages()
+{
+	MSG_Add("DOSBOX_HELP_LIST_COUNTRIES_1",
+	        "List of country codes (mostly same as telephone call codes)\n"
+	        "-----------------------------------------------------------");
+	MSG_Add("DOSBOX_HELP_LIST_COUNTRIES_2",
+	        "The above numeric country codes can be used exactly as listed\n"
+	        "in the 'country' conf setting.");
```
Comment (johnnovak) [inline src/dos/dos_locale.cpp:2897]: Maybe say "config setting".

File: src/dos/dos_locale.h [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+bool DOS_SetCountry(const uint16_t country_id);
+uint16_t DOS_GetCountry();
+void DOS_RefreshCountryInfo(const bool reason_keyboard_layout = false);
+
+std::string DOS_GetBundledCodePageFileName(const uint16_t code_page);
+uint16_t DOS_GetCodePageFromCountry(const uint16_t country);
+std::string DOS_CheckLanguageToLayoutException(const std::string& language_code);
+uint16_t DOS_GetDefaultCountry();
+bool DOS_GetCountryFromLayout(const std::string& layout, uint16_t& country);
+std::string DOS_GetLayoutFromHost();
```
Comment (johnnovak) [inline src/dos/dos_locale.h:43]: I like to break up such definitions a bit with blank lines, group them by function, etc. It's hard to read this block of text.

**Review by johnnovak — APPROVED** [review]
Another monstre effort from our resident localisation expert, @FeralChild64 ! 🎉  Just some minor comments, mostly about style and a few small improvements.

The special interest groups of DOS-enthusiast ex-acountants in Trinidad and Tobago, Antarctica, Zimbabwe, and Kyrgyzstan can finally relax—all currency symbols and decimal separators are now displayed in the correct format in their Lotus 1-2-3 spreadsheets! 😎 

Jokes aside, great job, man 😄 I admire your persistence, this must have been such a boring job 😅 We also need to make sure English is completely unaffected as that's the 99.9+% path at the end of the day.

File: src/dos/dos_locale.cpp [hunk 20]
```diff
@@ -0,0 +1,2904 @@
...
+constexpr size_t InfoOffsetListSeparator = 0x16;
+constexpr size_t InfoOffsetReserved      = 0x18;
+
+constexpr size_t MaxCurrencySymbolLength = 4;
+constexpr size_t ReservedAreaSize        = 10;
+
+static Country correct_country(const Country country)
+{
+	// Correct country code to handle duplicates in DOS country numbers
+	const auto country_int = static_cast<uint16_t>(country);
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: Done.

File: src/dos/dos_locale.cpp [hunk 21]
```diff
@@ -0,0 +1,2904 @@
...
+static void set_country_dos_code()
+{
+	if (config.country == Country::International) {
+		// MS-DOS uses the same country code for International English
+		// and Australia - we don't, as we have different settings for
+		// these. Let's imitate MS-DOS behavior.
+		config.country_dos_code = static_cast<uint16_t>(Country::Australia);
+	} else {
+		config.country_dos_code = static_cast<uint16_t>(config.country);
+	}
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp:1931]: Done.

File: src/dos/dos_locale.cpp (hunk 10 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: Done.

File: src/dos/dos_locale.cpp [hunk 22]
```diff
@@ -0,0 +1,2904 @@
...
+	return config.country_dos_code;
+}
+
+uint16_t DOS_GetDefaultCountry()
+{
+	if (config.locale_period == LocalePeriod::Historic) {
+		return static_cast<uint16_t>(Country::United_States);
+	} else {
+		return static_cast<uint16_t>(Country::International);
+	}
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp:2168]: Done.

File: src/dos/dos_locale.cpp [hunk 23]
```diff
@@ -0,0 +1,2904 @@
...
+uint16_t DOS_GetCodePageFromCountry(const uint16_t country)
+{
+	auto assert_code_page = [](const uint16_t code_page) {
+		assert(!DOS_GetBundledCodePageFileName(code_page).empty());
+		return code_page;
+	};
+
+	const auto country_id = static_cast<Country>(country);
+
+	// Countries grouped in ascending ordered by code_page value
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: Fixed.

File: src/dos/dos_locale.cpp (hunk 8 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: This is actually a quotation - reformatted the text.

File: src/dos/dos_locale.cpp [hunk 24]
```diff
@@ -0,0 +1,2904 @@
...
+	Semicolon  = 0x3b, // (;)
+};
+
+struct LocaleInfoEntry {
+	DosDateFormat date_format;
+	Separator     date_separator;
+	DosTimeFormat time_format;
+	Separator     time_separator;
+
+	// Several variants of currency symbol can be provided, ideally start
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, changed.

File: src/dos/dos_locale.cpp [hunk 25]
```diff
@@ -0,0 +1,2904 @@
...
+	// from "best, but few code pages can display it", to "worst, but many
+	// code pages can handle it".
+	// Add the rarely available symbols even for Historic locales; old code
+	// pages won't display them nevertheless and the engine will select
+	// more common one.
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, changed.

File: src/dos/dos_locale.cpp (hunk 9 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp:2016]: OK, changed.

**Comment by Grounded0** [comment]
I've been twice to Zimbabwe myself and i'm happy that the project serves our Zimbabwean users even better than before:

image

image

File: src/dos/dos_locale.h (hunk 19 above)
Comment (FeralChild64) [reply src/dos/dos_locale.h:48]: Ok, separated then - but I really have no idea how to group these,

File: docs/dosbox.1 [hunk 26]
```diff
@@ -89,6 +91,9 @@ Legacy single dash abbrevations still work for backwards compatibility.
 --working-dir <path>     Set working directory to <path>. DOSBox will act as if
                          started from this directory.
 
+--list-countries         List all supported countries with their numeric codes.
+                         Codes are useable in the 'country = ' config setting.
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply docs/dosbox.1]: OK, rephrased. I intended to have the same style help as in `--list-glshaders` description.

**Comment by FeralChild64** [comment]
> Another monstre effort

Few evenings, actually :D

File: src/dos/dos_locale.cpp (hunk 11 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, removed in several places, where it makes sense.

File: src/dos/dos_locale.cpp (hunk 13 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, split it into more functions.

File: src/dos/dos_locale.cpp [hunk 27]
```diff
@@ -0,0 +1,2904 @@
...
+	default: return {}; // none
+	}
+}
+
+DosDateFormat DOS_GetLocaleDateFormat()
+{
+	assert(dos.tables.country);
+
+	constexpr auto offset = InfoOffsetDateFormat;
+	return static_cast<DosDateFormat>(dos.tables.country[offset]);
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp:2774]: I am not aware about any.

File: src/dosbox.cpp [hunk 28]
```diff
@@ -1154,11 +1155,23 @@ void DOSBOX_Init()
...
-	pint = secprop->Add_int("country", when_idle, 0);
-	pint->Set_help("Set DOS country code (0 by default).\n"
-	               "This affects country-specific information such as date, time, and decimal\n"
-	               "formats. If set to 0, the country code corresponding to the selected keyboard\n"
-	               "layout will be used.");
+	secprop->AddInitFunction(&DOS_Locale_Init, changeable_at_runtime);
+	pstring = secprop->Add_string("locale_period", when_idle, "modern");
+	pstring->Set_help(
+	        "Set locale epoch ('modern' by default). Historic settings (if available\n"
+	        "for the given country) try to mimic old DOS behaviour when displaying\n"
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dosbox.cpp:1162]: Most often MS-DOS 6.22, sometimes Windows 9x, FreeDOS 1.3, or PC-DOS 2000. It is expected to mimic OS/2 for some countries in the future, where we have no DOS support.

I would prefer not to specify the exact revision.

File: src/dosbox.cpp [hunk 29]
```diff
@@ -1154,11 +1155,23 @@ void DOSBOX_Init()
...
+	        "practices for user experience more consistent with typical host systems.");
+	pstring->Set_values({"historic", "modern"});
+
+	pstring = secprop->Add_string("country", when_idle, "auto");
+	pstring->Set_help(
+	        "Set DOS country code ('auto' by default).\n"
+	        "This affects country-specific information such as date, time, and decimal\n"
+	        "formats. The list of supported country codes can be displayed using\n"
+	        "'--list-countries' command-line argument.\n"
+	        "If set to 'auto', the country code corresponding to the selected\n"
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dosbox.cpp]: OK, changed.

File: src/dos/dos_locale.cpp (hunk 18 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, changed.

File: src/dos/dos_locale.cpp [hunk 30]
```diff
@@ -0,0 +1,2904 @@
...
+static void auto_detect_country()
+{
+	const auto layout = DOS_GetLoadedLayout();
+	if (!layout) {
+		return;
+	}
+
+	uint16_t country_id = {};
+	if (!DOS_GetCountryFromLayout(layout, country_id)) {
+		LOG_ERR("DOS: The keyboard layout's country code: '%s' "
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: OK, changed.

File: src/dos/dos_keyboard_layout.cpp [hunk 31]
```diff
@@ -252,7 +251,7 @@ static uint32_t read_kcl_file(const FILE_unique_ptr &kcl_file, const char *layou
 			}
 		}
 		if (fseek(kcl_file.get(), cur_pos + 3 + len, SEEK_SET) != 0) {
-			LOG_ERR("LAYOUT: could not seek to byte %d in keyboard layout file: %s",
+			LOG_ERR("DOS: Could not seek to byte %d in keyboard layout file: %s",
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_keyboard_layout.cpp]: OK, simplified the log.

File: src/dos/dos_keyboard_layout.cpp [hunk 32]
```diff
@@ -319,8 +318,10 @@ KeyboardErrorCode KeyboardLayout::ReadKeyboardFile(const char *keyboard_file_nam
 		if (tempfile) {
 			const auto seek_pos = start_pos + 2;
 			if (fseek(tempfile.get(), seek_pos, SEEK_SET) != 0) {
-				LOG_WARNING("LAYOUT: could not seek to byte %d in keyboard layout file '%s': %s",
-				            seek_pos, keyboard_file_name, strerror(errno));
+				LOG_WARNING("DOS: Could not seek to byte %d in keyboard layout file '%s': %s",
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/dos_keyboard_layout.cpp]: OK, simplified the log.

File: src/dos/dos_locale.cpp (hunk 12 above)
Comment (FeralChild64) [reply src/dos/dos_locale.cpp]: IMHO - not worth it. Especially now, when this code went to a separate function.

File: src/dos/dos_keyboard_layout.cpp [hunk 33]
```diff
@@ -181,6 +178,12 @@ void KeyboardLayout::ReadKeyboardFile(const int32_t specific_layout)
 		                 dos.loaded_codepage);
 }
 
+static void log_layout_read_error()
+{
+	LOG_WARNING("DOS: Error reading keyboard layout file: '%s'",
+	            strerror(errno));
+}
+
```
Comment (kcgen) [inline src/dos/dos_keyboard_layout.cpp:186]: :+1: - nice cleanup with this function!

File: src/gui/sdlmain.cpp [hunk 34]
```diff
@@ -4556,6 +4563,14 @@ static int list_glshaders()
...
 }
 
+static int list_countries()
+{
+	const auto message_utf8 = DOS_GenerateListCountriesMessage();
+	printf_utf8("%s\n", message_utf8.c_str());
+
+	return 0;
+}
+
```
Comment (kcgen) [inline src/gui/sdlmain.cpp:4573]: The return code seems useless?

**Comment by johnnovak** [comment]
> I've been twice to Zimbabwe myself and i'm happy that the project serves our Zimbabwean users even better than before. Ndinokutendai mose!

Nice pics! 📷

**Comment by kcgen** [comment]
@johnnovak , DOS enthusiasts in Trinidad and Tobago 🇹🇹? You bet! 

https://dosprograms.info.tt/indexall.htm

(I grew up some years as a kid there, as my Dad was helping them with environment work during the oil boom of their coast in the '70s, so that website is extra nice to see! 😀)

**Review by kcgen — APPROVED** [review]
Another massive overhaul, @FeralChild64 ! 
(one very minor suggestion)

**Comment by johnnovak** [comment]
> @johnnovak , DOS enthusiasts in Trinidad and Tobago 🇹🇹?
> 
> You bet! https://dosprograms.info.tt/indexall.htm
> 
> (I grew up there for some years as a kid, as my Dad was helping them with environmental work during the oil boom off their coast in the '70s, so that website is extra nice to see! 😀)

Nice 😄 I picked a bad example because it seems like English is their official language, according to Wikipedia. I just wanted to pick a few countries I only know from the geography atlas 😆

**Comment by kcgen** [comment]
> Nice 😄 I picked a bad example because it seems like English is their official language, according to Wikipedia. 

I thought you might have known about their DOS pages 😅 ; pretty cool to see these lists still maintained. (Sometimes it comes up in searches for old DOS tools). Also very cool to see they've got several versions of DOSBox-X, too!

**Comment by johnnovak** [comment]
> > Nice 😄 I picked a bad example because it seems like English is their official language, according to Wikipedia.
> 
> I thought you might have known about their DOS pages 😅 ; pretty cool to see these lists still maintained. (Sometimes it comes up in searches for old DOS tools). Also very cool to see they've got several versions of DOSBox-X, too!

Yeah I came across that website but never looked into what the `.tt` domain stands for 😄 Nice website, yeah.

File: src/gui/sdlmain.cpp (hunk 34 above)
Comment (FeralChild64) [reply src/gui/sdlmain.cpp]: Fixed, also in `list_glshaders`.

**Pull request merged**